### PR TITLE
Basic scaler/reaper for petset

### PR DESCRIPTION
**Pull request by bprashanth**
Currently scaling or upgrading a petset is more complicated than it should be. Would be nice if this made code freeze on friday. I'm planning on a follow up change with generation number and e2es post freeze.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30813)

### CODE REVIEW

File: pkg/apis/apps/validation/validation.go [hunk 1]
```diff
@@ -99,23 +99,29 @@ func ValidatePetSet(petSet *apps.PetSet) field.ErrorList {
...
-	restoreGeneration := petSet.Generation
-	petSet.Generation = oldPetSet.Generation
+	restoreContainers := petSet.Spec.Template.Spec.Containers
+	petSet.Spec.Template.Spec.Containers = oldPetSet.Spec.Template.Spec.Containers
 
-	if !reflect.DeepEqual(petSet, oldPetSet) {
+	if !reflect.DeepEqual(petSet.Spec, oldPetSet.Spec) {
 		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec"), "updates to petset spec for fields other than 'replicas' are forbidden."))
 	}
+	if !reflect.DeepEqual(petSet.Status, oldPetSet.Status) {
```
Comment (smarterclayton) [inline pkg/apis/apps/validation/validation.go]: Shouldn't PrepareForUpdate be preventing these from differing?

File: pkg/kubectl/stop.go [hunk 2]
```diff
@@ -315,6 +324,53 @@ func (reaper *DaemonSetReaper) Stop(namespace, name string, timeout time.Duratio
...
+			if !errors.IsNotFound(err) {
+				errList = append(errList, err)
+			}
+		}
+	}
+	if len(errList) > 0 {
+		return utilerrors.NewAggregate(errList)
+	}
+
+	// TODO: Cleanup volumes? We don't want to accidentaly delete volumes from
```
Comment (smarterclayton) [inline pkg/kubectl/stop.go]: typo: accidentally

File: pkg/kubectl/stop.go [hunk 3]
```diff
@@ -315,6 +324,53 @@ func (reaper *DaemonSetReaper) Stop(namespace, name string, timeout time.Duratio
...
+	}
+	retry := NewRetryParams(reaper.pollInterval, reaper.timeout)
+	waitForPetSet := NewRetryParams(reaper.pollInterval, reaper.timeout)
+	if err = scaler.Scale(namespace, name, 0, nil, retry, waitForPetSet); err != nil {
+		return err
+	}
+
+	// TODO: This shouldn't be needed, see corresponding TODO in PetSetHasDesiredPets.
+	// PetSet should track generation number.
+	pods := reaper.Pods(namespace)
```
Comment (smarterclayton) [inline pkg/kubectl/stop.go]: Have pet sets been added to garbage collection yet?

Comment (bprashanth) [inline pkg/kubectl/stop.go]: Nope, only RC/RS AFAIK. I think we need generation and owner ref for gc support.

**Comment by smarterclayton** [comment]
Failing test.

**Comment by bprashanth** [comment]
Fixed unittest, ptal

**Comment by bprashanth** [comment]
(I would've inherited lgtm but noticed that Clayton added then removed lgtm, unsure if that was intentional)

**Comment by bprashanth** [comment]
Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/apis/apps/validation/validation.go, line 116 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30813#-KPPHF8boRcBazz95Dq1-r1-116:-KP_ZsWqNb3U5F73gfFf:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/a4886046d9177931fb25008b7ce2ef29b4777aab/pkg/apis/apps/validation/validation.go#L116)):*
<details><summary><i>Previously, smarterclayton (Clayton Coleman) wrote…</i></summary>
> Shouldn't PrepareForUpdate be preventing these from differing?
</details>
Done.

---

*[pkg/kubectl/stop.go, line 349 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30813#-KPPHF8cJDhcvj_P7Yqr-r1-349:-KP_ZtrfsH4pggvVq5ww:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/a4886046d9177931fb25008b7ce2ef29b4777aab/pkg/kubectl/stop.go#L349)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> Nope, only RC/RS AFAIK. I think we need generation and owner ref for gc support.
</details>
Done.

---

*[pkg/kubectl/stop.go, line 369 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30813#-KPPHF8cJDhcvj_P7Yqr-r1-369:-KP_ZtZeM_bjeljryLKx:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/a4886046d9177931fb25008b7ce2ef29b4777aab/pkg/kubectl/stop.go#L369)):*
<details><summary><i>Previously, smarterclayton (Clayton Coleman) wrote…</i></summary>
> typo: accidentally
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30813)*

**Comment by bprashanth** [comment]
hmm, i think this needs a reviewable LGTM + comment resolution for the presubmit hook to pass

**Comment by smarterclayton** [comment]
Hrm, tried to mark them resolved.  Reviewable sucks.

**Comment by smarterclayton** [comment]
:lgtm:

---

Reviewed 3 of 6 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30813#-:-KP_d4mffTTKfEh3N7ve:bnfp4nl)*

**Pull request merged**